web-registry/src/hooks/useQueryLedger.ts

```diff
@@ -0,0 +1,106 @@
+import { useState, useEffect } from 'react';
+
+import { QueryClientImpl as BankQueryClient } from '@regen-network/api/lib/generated/cosmos/bank/v1beta1/query';
+import { QueryClientImpl as EcocreditQueryClient } from '@regen-network/api/lib/generated/regen/ecocredit/v1alpha1/query';
+import { QueryClientImpl as BasketQueryClient } from '@regen-network/api/lib/generated/regen/ecocredit/basket/v1/query';
+
+import { useLedger } from '../ledger';
+
+// TODO: map clients and ledger queries here, hidding this details from business hooks
+// TODO: Map the clients and queries here, so that the implementation details
+// of the queries are transparent to business hooks. Instead, the hook would be called
+// like this:
+// ie. useQueryLedger('basket', { basketDenom })
+// ie. useQueryLedger('denomMetadata', { denom: basketDenom })
+
+type QueryClient = BankQueryClient | EcocreditQueryClient | BasketQueryClient;
+
+type LedgerModules = 'bank' | 'ecocredit' | 'basket';
+
+// TODO: generic type
+type LedgerQueryParams = any;
+
+// type LedgerQuery<P> = ({
+//   client,
+//   params,
+// }: {
+//   client: QueryClient;
+//   params: LedgerQueryParams<P>;
+// }) => Promise<any>;
+
+// type LedgerQuery = ({
+//   client,
+//   params,
+// }: {
+//   client: QueryClient;
+//   params: LedgerQueryParams;
+// }) => Promise<any>;
+
+// TODO: generic types
+type LedgerQuery = ({ client, params }: any) => Promise<any>;
+
+type InputQueryLedger = {
+  queryType: LedgerModules;
+  queryCallback: LedgerQuery;
+  queryParams?: LedgerQueryParams;
+};
+
+type OutputQueryLedger<T> = {
+  data: T | undefined;
+  loading: boolean;
+  error: Error | undefined;
+};
+
+export default function useQueryLedger<T>({
+  queryType,
+  queryCallback,
+  queryParams,
```

Comment:
we could just use `type`, `callback` and `params`, using this `query` prefix is a bit redundant.

also this design doesn't refrain the developer from specifying a query type like 'bank' and then trying to call a query callback that is not related to bank

Reply:
yes, totally... I also thought about changing it at some point. I was also about to encapsulate all those details in the useQueryLedger hook itself so that you only have to specify the query name and query parameters.